src/lib/Type.hs

```diff
@@ -508,34 +543,56 @@ typeCheckTyCon :: TC -> TypeM Type
 typeCheckTyCon tc = case tc of
   BaseType _       -> return TyKind
   CharType         -> return TyKind
-  ArrayType t      -> t|:TyKind >> return TyKind
   IntRange a b     -> a|:IdxRepTy >> b|:IdxRepTy >> return TyKind
   IndexRange t a b -> t|:TyKind >> mapM_ (|:t) a >> mapM_ (|:t) b >> return TyKind
   IndexSlice n l   -> n|:TyKind >> l|:TyKind >> return TyKind
   PairType a b     -> a|:TyKind >> b|:TyKind >> return TyKind
   UnitType         -> return TyKind
-  RefType r a      -> r|:TyKind >> a|:TyKind >> return TyKind
+  RefType (Just r) a  -> r|:TyKind >> a|:TyKind >> return TyKind
+  RefType Nothing a   ->              a|:TyKind >> return TyKind
   TypeKind         -> return TyKind
   EffectRowKind    -> return TyKind
   LabeledRowKindTC -> return TyKind
-  JArrayType _ _   -> undefined
 
 typeCheckCon :: Con -> TypeM Type
 typeCheckCon con = case con of
   Lit l -> return $ BaseTy $ litType l
   CharCon v -> v |: (BaseTy $ Scalar Int8Type) $> CharTy
-  ArrayLit ty _ -> return $ ArrayTy ty
+  -- BufferLit buffer -> return $ BaseTy $ snd $ arrayType buffer
   AnyValue t -> t|:TyKind $> t
   PairCon x y -> PairTy <$> typeCheck x <*> typeCheck y
   UnitCon -> return UnitTy
-  Coerce t@(Var _) _ -> t |: TyKind $> t
-  Coerce destTy e -> do
-    sourceTy <- typeCheck e
-    destTy  |: TyKind
-    checkValidCoercion sourceTy destTy
-    return destTy
-  SumAsProd ty _ _ -> return ty  -- TODO: check!
+  SumAsProd ty tag _ -> tag |:TagRepTy >> return ty  -- TODO: check!
   ClassDictHole _ ty -> ty |: TyKind >> return ty
+  IntRangeVal     l h i -> i|:IdxRepTy >> return (TC $ IntRange     l h)
+  IndexRangeVal t l h i -> i|:IdxRepTy >> return (TC $ IndexRange t l h)
+  BaseTypeRef p -> do
+    (PtrTy (_, _, b)) <- typeCheck p
+    return $ RawRefTy $ BaseTy b
+  TabRef tabTy -> do
+    TabTy b (RawRefTy a) <- typeCheck tabTy
+    return $ RawRefTy $ TabTy b a
+  ConRef con -> case con of
```

Comment:
It's a little weird that we have both a `ConRef` and `BaseTypeRef` and `TabRef`. Couldn't we put all of those under `ConRef`?

Reply:
`TabRef` might have to stay but we could get rid of `BaseTypeRef`